Add `ShootState` controller managing the object deletion with additional `finalizer` (#11491)
gardener/gardener: docs/concepts/controller-manager.md

A `Shoot` is considered ready for migration if the destination `Seed` is up-to-date and healthy.

The main purpose of this constraint is to allow the `gardenlet` running in the source seed cluster to check if it can start with the migration flow without that it needs to directly read the destination `Seed` resource (for which it won't have permissions).

#### ["ShootState Finalizer" Reconciler](../../pkg/controllermanager/controller/shoot/state/finalizer)

This reconciler is responsible for managing a `ShootState` finalizer that
ensures the object existence during migration of `Shoot`s control plane
to another `Seed`.

The main goal is to keep the `ShootState` present during the `Migrate` and

> `Restore` operations that are not yet finished successfully.

Reviewer: one line per sentence please

Author: Thanks for pointing it, fixed in the latest updates.

A `Shoot` is considered ready for migration if the destination `Seed` is up-to-date and healthy.

The main purpose of this constraint is to allow the `gardenlet` running in the source seed cluster to check if it can start with the migration flow without that it needs to directly read the destination `Seed` resource (for which it won't have permissions).

> #### ["ShootState Finalizer" Reconciler](../../pkg/controllermanager/controller/shoot/state/finalizer)

Reviewer 2: ```suggestion
#### ["ShootState Finalizer" Reconciler](../../pkg/controllermanager/controller/shootstate)
```

Author: Thanks for catching this directory mismatch. Fixed it in my latest updates.

> This reconciler is responsible for managing a `ShootState` finalizer that ensures the object existence during migration of `Shoot`s control plane to another `Seed`.

Reviewer 2: A wording suggestion. Letting you to decided whether you want to accept it or not:
```suggestion
This reconciler is responsible for managing a finalizer (`core.gardener.cloud/shootstate`) on a `ShootState`. The finalizer ensures the `ShootState` will exist during migration of `Shoot`'s control plane to another `Seed`.
```

Author: Thanks for the suggestion, it definitely adds more context to the matter. Reused it in my latest updates.

> The main goal is to keep the `ShootState` present during the `Migrate` and `Restore` operations that are not yet finished successfully.

Reviewer 2: A wording suggestion. Letting you to decided whether you want to accept it or not:

```suggestion
The `ShootState` has to be present until the `Migrate` and `Restore` operations that are not yet finished successfully. Otherwise, in some corner cases the `ShootState` could be deleted before the `Restore` operations succeeds and when the `Restore` operation is retried, it fails because of already deleted `ShootState` resource.
```

Author: Thanks for the suggestion, I reused part of it with some slight adjustments to the last portion, but preserving the meaning.

```md
The `ShootState` has to be present until the `Migrate` and `Restore` operations finish successfully. Otherwise, in corner cases of prior deletion, subsequent `Restore` operations of the `Shoot` will fail due to the missing `ShootState` resource.
```
